pkg/patroni/failover.go

```diff
@@ -0,0 +1,199 @@
+}
+
+// CheckForFailover checks if a failover has occurred since last check.
+// Returns FailoverDetectedError if primary changed, nil otherwise.
+func (adapter *PatroniAdapter) CheckForFailover(ctx context.Context) error {
+	logger := adapter.Logger()
+
+	// Get cluster status from Patroni REST API
+	clusterStatus, err := adapter.getPatroniClusterStatus(ctx)
+	lastKnownPrimary := adapter.State.GetLastKnownPrimary()
+
+	// If we can't get cluster status AND we previously had a primary tracked,
+	// this likely means failover is in progress
+	if err != nil {
+		if lastKnownPrimary != "" {
+			logger.Warnf("Cannot get Patroni cluster status (failover likely in progress): %v", err)
+			return &FailoverDetectedError{
+				OldPrimary: lastKnownPrimary,
+				NewPrimary: "(unknown - failover in progress)",
+				Message:    fmt.Sprintf("cluster status unavailable: %v", err),
+			}
+		}
+		// First check and can't get status - just log and skip
+		logger.Warnf("Failed to get Patroni cluster status for initial check: %v", err)
+		return nil
+	}
+
+	currentPrimary := clusterStatus.CurrentPrimary
+
+	// If cluster reports no current primary AND we had one before, failover is happening
+	if currentPrimary == "" && lastKnownPrimary != "" {
+		logger.Warnf("Patroni cluster has no current primary (failover in progress, was: %s)", lastKnownPrimary)
+		return &FailoverDetectedError{
+			OldPrimary: lastKnownPrimary,
+			NewPrimary: "(none - failover in progress)",
+			Message:    fmt.Sprintf("cluster state: %s", clusterStatus.State),
+		}
+	}
+
+	// First time checking - just record current primary
+	if lastKnownPrimary == "" {
+		if currentPrimary != "" {
+			adapter.State.SetLastKnownPrimary(currentPrimary)
+			logger.Infof("Tracking primary node: %s", currentPrimary)
+		}
+		return nil
+	}
+
+	// Check if primary changed (failover occurred)
+	if currentPrimary != lastKnownPrimary {
+		logger.Warnf("Failover detected: %s → %s (state: %s)",
+			lastKnownPrimary, currentPrimary, clusterStatus.State)
+
+		return &FailoverDetectedError{
+			OldPrimary: lastKnownPrimary,
+			NewPrimary: currentPrimary,
+			Message:    fmt.Sprintf("cluster state: %s", clusterStatus.State),
+		}
+	}
+
+	return nil
+}
+
+// HandleFailoverDetected is called when a failover is detected.
+// It sends an error to the backend and updates tracking to the new primary.
+func (adapter *PatroniAdapter) HandleFailoverDetected(ctx context.Context, failoverErr *FailoverDetectedError) error {
+	logger := adapter.Logger()
+
+	logger.Errorf("Failover detected: %s", failoverErr.Error())
+	logger.Info("Sending failover notification to backend")
+
+	// Send failover error to backend
+	// Backend decides: stop tuning session? apply baseline? etc.
+	errorPayload := agent.ErrorPayload{
+		ErrorMessage: failoverErr.Error(),
+		ErrorType:    "failover_detected",
+		Timestamp:    time.Now().UTC().Format(time.RFC3339),
+	}
+	adapter.SendError(errorPayload)
```

Comment:
This is good, just that if we want to support restarts, we don't want to send this everytime.